crates/conjure_core/src/rule_engine/rewrite_naive.rs

```diff
@@ -89,11 +89,24 @@ pub fn rewrite_naive<'a>(
                 result.reduction.clone().apply(&mut model);
 
                 if results.len() > 1 && prop_multiple_equally_applicable {
-                    let names: Vec<_> = results
```

Comment:
I am not getting this part. Due to how we iterate over the rules in priority order and collect all applicable rules at a level, I don't think you will ever have results originating from different priority levels at this point in the code. Am I missing something?

Reply:
in fact it is not checking whether lower priority rules are also applicable at all.

Reply:
> in fact it is not checking whether lower priority rules are also applicable at all.

on main i get the following error:

```
multiple equally applicable rules for (x) -> (x): [
    "partial_evaluator",
    "introduce_reifyimply_ineq_from_imply",
]
```

Reply:
looks like we never check if lower priority rule sets apply, but we do exhaustively check all rules inside a ruleset.

Reply:
In results, we keep track of all the highest applicable rules *per expression*. So we can have multiple rules of different priorities in there targeting different expressions.

Reply:
I think that needs a separate fix! Line 73 should break out of both for loops! Right?

A case for labelled breaks?

Reply:
the order of the loops should guarantee that the first time `results` has something in it, that is the highest priority applicable rule.

we first consider the highest priority ruleset, and are ready to search the universe to find an expression to apply one of them. we only try lower priority expressions if we fail to apply the highest priority ruleset. and so on for the remainder of the priority levels.

unless we are not doing that, but I think we are.

Reply:
> the order of the loops should guarantee that the first time `results` has something in it, that is the highest priority applicable rule.

in fact, even stronger: the first time `results` has something in it, it will contain all highest priority rule applications to the same expression (in `contexts_bi` order)

Reply:
> I think that needs a separate fix! Line 73 should break out of both for loops! Right?
> 
> A case for labelled breaks?

Looking deeper into this, it looks like our tie-breaker is not "shallowest" expression, but rather which one comes first in a preorder traversal. Shallowest expression seems expensive to do, so I am just going to update the docs to reflect reality.